### PR TITLE
Fix jemalloc and qemu tests on nixos

### PR DESCRIPTION
**Pull request by fidgetingbits**
Fix the nix dev shell environment to allow running jemalloc tests, as well as qemu tests. I also removed `with pkgs` (as per latest nix language recommendations). @patryk4815 to review the nix bits.

While testing jemalloc I ran into a few edge cases, so also added some minor tweaks to better handle errors. @jetchirag to review the jemalloc bits.

### CODE REVIEW

**Comment by fidgetingbits** [comment]
I guess the gdb memory errors I throw will break things due to lldb, so will fix.

**Comment by patryk4815** [comment]
@fidgetingbits  I think it is safe to `raise pwndbg.dbg_mod.Error('msg')`

**Comment by fidgetingbits** [comment]
Thanks @patryk4815. The error throwing should be fixed to be agnostic now. How do I run the tests in lldb?

Although the tests do run, I notice that I get a failed test:

```bash
        for i in range(len(expected_output)):
>           assert re.match(expected_output[i], result[i])
E           AssertionError: assert None
E            +  where None = <function match at 0x7f9c231c5c60>('Allocated Address: 0x7ffff[0-9a-fA-F]{6,9}', 'Allocated Address: 0x1c')
E            +    where <function match at 0x7f9c231c5c60> = re.match

EXPECTED_EXTENT_ADDRESS = 140737347937664
expected_output = ['Jemalloc extent info',
 'This command was tested only for jemalloc 5.3.0 and does not support lower '
 'versions',
 '',
 'Allocated Address: 0x7ffff[0-9a-fA-F]{6,9}',
 'Extent Address: 0x7ffff[0-9a-fA-F]{6,9}',
 'Size: 0x1000',
 'Small class: True']
i          = 3
result     = ['Jemalloc extent info',
 'This command was tested only for jemalloc 5.3.0 and does not support lower '
 'versions',
 '',
 'Allocated Address: 0x1c',
 'Extent Address: 0x7ffff7a16580',
 'Size: 0x11001200000000',
 'Small class: False',
 'State: Muzzy']
start_binary = <function start_binary.<locals>._start_binary at 0x7f9c0fd982c0>
```

This seems to be because it's using a hardcoded address 0x7FFFF7A16580 which won't be valid on most systems I guess. Should just pull out the extent address dynamically from find_extent output imo. This can likely be fixed as part of another PR though.

**Comment by patryk4815** [comment]
@fidgetingbits My last PR probably messed up those tests. I asked disconnect3d to investigate, but instead he merged the PR xD

**Comment by patryk4815** [comment]
@fidgetingbits only one test failing? this one test with hardcoded `0x7FFFF7A16580` looks wrong

**Comment by fidgetingbits** [comment]
Are these mypy errors related to aglib expected atm?

**Comment by fidgetingbits** [comment]
> @fidgetingbits only one test failing? this one test with hardcoded `0x7FFFF7A16580` looks wrong

```bash
❯ ./tests.sh --nix jemalloc
ZIGPATH set to /nix/store/wwcqxrjj91lcmdp3h8b3a2dghpq9i04s-zig-0.10.1/bin/
make: Nothing to be done for 'all'.

Running tests in parallel
test_jemalloc_heap                                                     SKIPPED
test_jemalloc_find_extent                                              PASSED
test_jemalloc_extent_info                                              FAILED
```

The hardcoded `0x7FFFF7A16580` test is the one I mentioned above. It should be easy to fix. I can just do it as part of this PR I guess.

**Comment by fidgetingbits** [comment]
Fixed. Don't plan to investigate why the other test is skipped atm

```bash
./tests.sh --nix jemalloc
ZIGPATH set to /nix/store/wwcqxrjj91lcmdp3h8b3a2dghpq9i04s-zig-0.10.1/bin/
make: Nothing to be done for 'all'.

Running tests in parallel
test_jemalloc_heap                                                     SKIPPED
test_jemalloc_find_extent                                              PASSED
test_jemalloc_extent_info                                              PASSED
Tests completed in 1 seconds

*********************************
********* TESTS SUMMARY *********
*********************************
Tests Passed: 2
Tests Skipped: 1
Tests Failed: 0
```

**Comment by fidgetingbits** [comment]
Hrm, so the CI is failing on the test that's skipped in my environment, so maybe I'll look into it later:

```bash
test_commands[jemalloc_heap]                                           XFAIL

pwndbg: loaded 173 pwndbg commands and 47 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $base, $bn_sym, $bn_var, $bn_eval, $ida GDB functions (can be used with print/break)
['/home/runner/work/pwndbg/pwndbg', '/home/runner/work/pwndbg/pwndbg/.venv/lib/python3.10/site-packages', '/usr/share/gdb/python', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/home/runner/work/pwndbg/pwndbg']
Launching pytest with args: ['/home/runner/work/pwndbg/pwndbg/tests/gdb-tests/tests/test_commands.py::test_commands[jemalloc_heap]', '-vvv', '-s', '--showlocals', '--color=yes']
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.0.2, pluggy-1.5.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/pwndbg/pwndbg/tests
plugins: cov-4.1.0
collecting ... collected 1 item

gdb-tests/tests/test_commands.py::test_commands[jemalloc_heap] Running command jemalloc_heap

Program stopped.
0x00007ffff7fe3290 in _start () from /lib64/ld-linux-x86-64.so.2
Jemalloc heap
This command was tested only for jemalloc 5.3.0 and does not support lower versions

XFAIL

============================== 1 xfailed in 0.20s ==============================
```

**Comment by patryk4815** [comment]
`jemalloc_heap` test is failing on current master/dev branch

**Comment by patryk4815** [comment]
If you want fix `test_commands[jemalloc_heap]` test. 
Replace `jemalloc_heap` func with checking exception from `jemalloc.RTree.get_rtree`
```
@pwndbg.commands.ArgparsedCommand(parser, category=CommandCategory.HEAP)
def jemalloc_heap() -> None:
    print(C.banner("Jemalloc heap"))
    print("This command was tested only for jemalloc 5.3.0 and does not support lower versions")
    print()

    try:
        rtree = jemalloc.RTree.get_rtree()
    except pwndbg.dbg_mod.Error as e:
        print(message.warn(f"{e}"))
        return
```

**Comment by patryk4815** [comment]
(we don't have unittests yet)
If you want test lldb:
```command
$ nix develop
$ make -C ./tests/gdb-tests/tests/binaries/ heap_jemalloc_extent_info.out
$ nix build .#pwndbg-lldb -o result-lldb
$ ./result-lldb/bin/pwndbg-lldb ./tests/gdb-tests/tests/binaries/heap_jemalloc_extent_info.out
pwndbg-lldb> b break_here
pwndbg-lldb> r
pwndbg-lldb> jemalloc_find_extent ptr
```

**Review by disconnect3d — APPROVED** [review]

File: pwndbg/aglib/heap/jemalloc.py [hunk 1]
```diff
@@ -203,13 +207,10 @@ def __init__(self, addr: int) -> None:
...
     @staticmethod
-    def get_rtree() -> RTree | None:
-        try:
-            addr = pwndbg.dbg.selected_inferior().symbol_address_from_name("je_arena_emap_global")
-            if addr is None:
-                return None
-        except pwndbg.dbg_mod.Error:
-            return None
+    def get_rtree() -> RTree:
+        addr = pwndbg.aglib.info.address("je_arena_emap_global")
```
Comment (disconnect3d) [inline pwndbg/aglib/heap/jemalloc.py]:
```
=================================== FAILURES ===================================
__________________________ test_jemalloc_find_extent ___________________________

start_binary = <function start_binary.<locals>._start_binary at 0x7efdf53ca680>

    def test_jemalloc_find_extent(start_binary):
        start_binary(HEAP_JEMALLOC_EXTENT_INFO)
        gdb.execute("break break_here")
        gdb.execute("continue")
    
        # run jemalloc extent_info command
>       result = gdb.execute("jemalloc_find_extent ptr", to_string=True).splitlines()
E       gdb.error: Error occurred in Python: module 'pwndbg.aglib' has no attribute 'info'

start_binary = <function start_binary.<locals>._start_binary at 0x7efdf53ca680>
```

**Review by disconnect3d — CHANGES_REQUESTED** [review]
a test is failing due to the changes here right now

**Comment by fidgetingbits** [comment]
> a test is failing due to the changes here right now

will look into it

**Comment by fidgetingbits** [comment]
@disconnect3d identified test regression should be fixed. I haven't looked into the other test, and I think it's better to just address that in a separate issue/PR.

As to the fix I just added for this test regression: having to use `pwndbg.dbg.selected_inferior().symbol_address_from_name("je_arena_emap_global")` is pretty gross. Perhaps aglib should add a function to make that much more ergonomic? If there isn't already an issue filed for that I can make one.

Also not exactly easy to find out what the "right" function that works with aglib is (been awhile since I contributed, so I think aglib is new to me). Makes me think aglib could use a translation documentation to show if you normally use `gdb.foo` then you should instead use `aglib.bar`.

**Comment by fidgetingbits** [comment]
Ok I lied. I took a quick look at the other test and fixed it, but probably worth someone else taking a quick look to make sure it's sane. Seems like my extent output is different than what @jetchirag had, but the first extent size is the same. Not sure if this has to do with a slightly different build environment. I've made the test a bit more forgiving, since it at least shows that the command is working (in that it's dumping about extents), which is probably fine for now.

**Comment by disconnect3d** [comment]
> As to the fix I just added for this test regression: having to use pwndbg.dbg.selected_inferior().symbol_address_from_name("je_arena_emap_global") is pretty gross. Perhaps aglib should add a function to make that much more ergonomic? If there isn't already an issue filed for that I can make one.

> Also not exactly easy to find out what the "right" function that works with aglib is (been awhile since I contributed, so I think aglib is new to me). Makes me think aglib could use a translation documentation to show if you normally use gdb.foo then you should instead use aglib.bar.

I agree with both of those points @fidgetingbits (cc: @mbrla0)

**Comment by disconnect3d** [comment]
@fidgetingbits this now conflicts with DEVELOPING.md :<

**Comment by fidgetingbits** [comment]
> @fidgetingbits this now conflicts with DEVELOPING.md :<

Fixed. 

It's not clear from the CI log why the unit tests are failing, so not sure if I should just re-disable it in this PR or what. Thoughts?

**Comment by disconnect3d** [comment]
> > @fidgetingbits this now conflicts with DEVELOPING.md :<
> 
> Fixed.
> 
> It's not clear from the CI log why the unit tests are failing, so not sure if I should just re-disable it in this PR or what. Thoughts?

Ugh
```
*********************************
********* TESTS SUMMARY *********
*********************************
Tests Passed: 420
Tests Skipped: 1
Tests Failed: 2

Failing tests: 

```
yeah this display is suboptimal. The failed tests are the "XFAIL" tests from "test_commands.py". Its `bugreport`  and `jemalloc_<something>` commands failing when no args are passed to them. The display in CI doesnt show the exact bug logs for them which sucks.

This can be debugged further by modifying test_commands.py to filter just one of those commands, then running `./tests.py test_command --pdb` and getting a proper stacktrace for them.

At least thats how I fixed a couple other bugs like that we had discovered recently. We need to fix the other two as well as fix the display so that:
1. "Failing tests:" actually show the "XFAIL" failed  tests too
2. The test_command displays proper stacktraces for failures

**Comment by disconnect3d** [comment]
Anyway, those failures are unrelated to this PR so it can be merged

**Review by disconnect3d — APPROVED** [review]

**Comment by fidgetingbits** [comment]
> > > @fidgetingbits this now conflicts with DEVELOPING.md :<
> > 
> > 
> > Fixed.
> > It's not clear from the CI log why the unit tests are failing, so not sure if I should just re-disable it in this PR or what. Thoughts?
> 
> Ugh
> 
> ```
> *********************************
> ********* TESTS SUMMARY *********
> *********************************
> Tests Passed: 420
> Tests Skipped: 1
> Tests Failed: 2
> 
> Failing tests: 
> ```
> 
> yeah this display is suboptimal. The failed tests are the "XFAIL" tests from "test_commands.py". Its `bugreport` and `jemalloc_<something>` commands failing when no args are passed to them. The display in CI doesnt show the exact bug logs for them which sucks.
> 
> This can be debugged further by modifying test_commands.py to filter just one of those commands, then running `./tests.py test_command --pdb` and getting a proper stacktrace for them.
> 
> At least thats how I fixed a couple other bugs like that we had discovered recently. We need to fix the other two as well as fix the display so that:
> 
>     1. "Failing tests:" actually show the "XFAIL" failed  tests too
> 
>     2. The test_command displays proper stacktraces for failures

Ah, thanks I understand now. The `test_commands.py` script is running it on `heap_bins.out` which won't have the jemalloc symbols, and atm the code just raises an exception. So I guess all the jemalloc commands will have to be updated to more gracefully handle exceptions and print an error. @jetchirag 

For some reason I was incorrectly reading it as one of the actual jemalloc tests failing. Now I understand why @patryk4815 suggested just catching the exceptions lol.
